### DISCUSSION TITLE
Type pool

### DISCUSSION THREAD

**Discussion by weirdan** [vimeo/psalm#9328] [Psalm development (AKA internals)]
### Problem statement

Psalm, naturally, operate on a lot of types. My intuition (and some preliminary testing) suggests that we have a lot of duplicate type instances, maybe about ten times more than strictly necessary. This leads to Psalm having to use more memory and prevents some performance optimizations. 

### Where do duplicates come from?

One example could be literally duplicate types used by the inspected codebase. E.g., how many instances of `TTrue` do we have during analysis? And how many do we actually need? I'd say we should ideally have just two - one with `$from_docblock = true` and one with `$from_docblock = false`.

### Idea

Instead of having multiple instances of identical types, we could have a central pool of type instances and return instances from that pool instead of creating new instances.

### Possible optimizations it enables

Comparing types could be short-circuited to identity comparison. If there's only a single instance of a non-docblock `TArray<array-key, mixed>` we could first check `if ($input_type === $container_type) { return true; }` and save some CPU cycles that are otherwise wasted comparing key and value types.

### Why can't we do this already?

Several architectural issues are preventing the immediate implementation of this approach.

1. Types do have transient state. E.g. `Atomic::$checked`
2. Types carry source information. E.g. `Atomic::$offset_start` & `Atomic::$offset_end`

There's also a question of how this should work with the cache. We would probably need to serialize the type pool into the cache and replace all type references with some kind of pointer to the serialized pool.

### What would the code look like if we implemented it?

Parts of the code that create new types would need to register the created type with the pool, and let it replace the type instance. E.g. parser would do this:

```php
public function parse(string $type_string): Union 
{
  // ... do the parsing resulting in $type being created
  // register the type with the pool, potentially receiving another instance with identical content
  $type = TypePool::register($type);
  // return the type
  return $type;
}
```

@orklah @danog @muglug do you think this is feasible and actually makes sense?

**Comment by danog** [comment]
I like the idea a lot.
Ever since implementing immutability, I saw this as a chance to reuse at least some types; I actually played around with the idea a bit, implementing memoization for all type getter methods in https://github.com/vimeo/psalm/blob/master/src/Psalm/Type.php (with from_docblock distinction); this didn't yield a noticeable performance improvement, but it did reduce memory usage.

Extending the idea with a full union and atomic type cache, including composite types like arrays and simplifying comparisons should improve performance a lot more IMO.

**Comment by orklah** [comment]
If we manage to extract dynamic parts from Atomic, it may make sense to make something like singletons for simple type. Atomic/Union creation are pretty centralised already so it would not be a big change to make the constructor private and just rely on the single instance available. That would make the need for a whole pool system unnecessary

**Comment by edsrzf** [comment]
This makes a lot of sense to me. Is it something I can help with? (Keeping in mind I'm still pretty new to the code base.)

I've started looking at `Atomic::$offset_start` and `Atomic::$offset_end`, and it looks like they really only need to be present on `TNamedObject`, here: https://github.com/vimeo/psalm/blob/c0e757cfa9c266dafb16c41f3f96c8fac374b319/src/Psalm/Internal/TypeVisitor/TypeChecker.php#L140-L141

Intuitively this kinda makes sense to me, since you typically wouldn't ask questions like "Where are all the places in the code base I'm using `bool`?" You'd only care about named types that you've defined.

To make the most of a type pool, I suppose you'd want to eliminate offset info even on these named type objects, but it seems to me there'd be a lot of benefit in covering the built-in types by themselves.

I guess this work would need to target Psalm 6, since it would almost certainly contain breaking changes for plugins?

**Reply by weirdan** [reply]
> I guess this work would need to target Psalm 6, since it would almost certainly contain breaking changes for plugins?

Absolutely. We can start by marking those properties `@deprecated` now and work with those plugin authors who would complain (if any). 

> I've started looking at Atomic::$offset_start and Atomic::$offset_end

There's also `$text`, but also an assortment of properties on various descendant classes.

>  Is it something I can help with?

A PR that makes all atomics consistently call parent constructors would be welcome. Once that is done, we can see how many atomics we instantiate (e.g. using a profiler).

**Reply by edsrzf** [reply]
I opened #9406. In my local copy I added some basic tracking and ran Psalm over itself and over the closed-source code base of a web application I work on.

I've put some details below of the breakdown of constructed atomics, for types that made up 1% or more of the constructed Atomics. Note that I didn't instrument clones and I'm now realizing I probably should have. I can try another run tomorrow.

I'm a little surprised that TLiteralString is so common in both code bases. If I track the values, it looks like most of them are only used once.

## Psalm Code Base

836,073 Atomics constructed.

| Type | Percent |
|-|-|
| TLiteralString | 13.3% |
| TKeyedArray | 11.1% |
| TNever | 9.3% |
| TArray | 9.0% |
| TBool | 7.9% |
| TString | 7.8% |
| TNamedObject | 7.6% |
| TLiteralInt | 4.2% |
| TNonEmptyString | 3.1% |
| TIntRange | 3.1% |
| TInt | 3.0% |
| TNonEmptyArray | 3.0% |
| TNull | 2.7% |
| TMixed | 2.6% |
| TFalse | 2.1% |
| TTrue | 1.8% |
| TLowercaseString | 1.3% |
| TFloat | 1.1% |
| TNumericString | 1.1% |

## Closed-Source Code Base

3,738,177 Atomics constructed.

| Type | Percent |
|-|-|
| TMixed | 18.8% |
| TLiteralString | 18.2% |
| TBool | 8.6% |
| TNever | 7.3% |
| TArray | 6.3% |
| TKeyedArray | 5.8% |
| TNamedObject | 5.8% |
| TNonEmptyArray | 4.1% |
| TString | 3.8% |
| TNull | 2.8% |
| TLiteralInt | 2.7% |
| TConditional | 2.0% |
| TIntRange | 2.0% |
| TFalse | 1.7% |
| TArrayKey | 1.3% |
| TInt | 1.1% |
| TTrue | 1.0% |

**Reply by edsrzf** [reply]
Same analysis, with clones included this time. TArray has really started catching up in both. TTemplateParam didn't even show up in the first analysis because it was below 1%, but here it makes a solid appearance in both code bases.

## Psalm Code Base

Psalm is up to 1,034,740 Atomic allocations and TArray takes the lead.

| Type | Percent |
|-|-|
| TArray | 16.3% |
| TLiteralString | 11.1% |
| TKeyedArray | 9.6% |
| TNever | 7.6% |
| TNamedObject | 7.1% |
| TString | 7.0% |
| TBool | 6.4% |
| TLiteralInt | 3.6% |
| TTemplateParam | 3.5% |
| TNonEmptyArray | 3.3% |
| TIntRange | 2.8% |
| TNonEmptyString | 2.6% |
| TNull | 2.6% |
| TInt | 2.6% |
| TConditional | 2.4% |
| TMixed | 2.1% |
| TFalse | 1.7% |
| TTrue | 1.5% |
| TLowercaseString | 1.0% |

## Closed-Source Code Base

We're up to 5,125,195 allocations, with TMixed still solidly in front. TArray is catching up to TLiteralString. I'm very surprised that TTemplateParam is so high because this code base hardly uses any templated types.

| Type | Percent |
|-|-|
| TMixed | 17.8% |
| TLiteralString | 13.7% |
| TArray | 12.1% |
| TTemplateParam | 7.3% |
| TBool | 6.3% |
| TNever | 5.3% |
| TKeyedArray | 5.3% |
| TNamedObject | 5% |
| TConditional | 4.5% |
| TNonEmptyArray | 3.9% |
| TString | 3.1% |
| TNull | 2.6% |
| TLiteralInt | 2.2% |
| TIntRange | 1.6% |
| TFalse | 1.3% |
| TArrayKey | 1.2% |

**Reply by weirdan** [reply]
> Psalm is up to 1,034,740 Atomic allocations 

This sounds a bit too low. Did you run it with `--threads=1`?

My backfire profile shows ~4M calls to `Union::__construct()` alone, and I'd expect to see at least that number of `Atomic` instances.

**Reply by weirdan** [reply]
>  Did you run it with --threads=1?

With multiprocessing (`--threads` set to anything higher, and it's >1 by default) analysis happens in child processes, so you may miss all the allocations that happen there.

**Reply by edsrzf** [reply]
Yes, I ran with `--threads=1`.

**Reply by edsrzf** [reply]
Is it possible that I'm missing some allocations through `unserialize`? I also ran with `--no-cache`, but maybe there's still some (de)serializing happening somewhere?
